chi/hardware.py

```diff
@@ -0,0 +1,155 @@
+
+import requests
+import logging
+
+LOG = logging.getLogger(__name__)
+
+@dataclass
+class Node:
+    """
+    Represents the Chameleon hardware that goes into a single node.
+    A dataclass for node information directly from the hardware browser.
+    """
+    site: str
+    name: str
+    type: str
+    architecture: dict
+    bios: dict
+    cpu: dict
+    gpu: dict
+    main_memory: dict
+    network_adapters: List[dict]
+    placement: dict
+    storage_devices: List[dict]
+    uid: str
+    version: str
+
+    def next_free_timeslot(self) -> datetime:
+        """
+        (Not implemented yet) Finds the next available timeslot for the hardware.
+
+        Returns:
+            A tuple containing the start and end datetime of the next available timeslot.
+            If no timeslot is available, returns the end datetime of the last lease.
+
+        """
+        raise NotImplementedError()
+
+        blazarclient = blazar()
+
+        # Get all leases
+        leases = blazarclient.lease.list()
+
+        # Filter leases for this node
+        node_leases = [
+            lease for lease in leases
+            if any(r['resource_id'] == self.uid for r in lease.get('reservations', [])
+                   if r['resource_type'] == 'physical:host')
+        ]
+
+        # Sort leases by start time
+        node_leases.sort(key=lambda x: x['start_date'])
+
+        now = datetime.now(timezone.utc)
+
+        print(node_leases)
+
+        # Check if there's a free slot now
+        if not node_leases or node_leases[0]['start_date'] > now:
+            return (now, node_leases[0]['start_date'] if node_leases else None)
+
+        # Find the next free slot
+        for i in range(len(node_leases) - 1):
+            current_end = datetime.strptime(node_leases[i]['end_date'], "%Y-%m-%d %H:%M:%S")
+            next_start = datetime.strptime(node_leases[i+1]['start_date'], "%Y-%m-%d %H:%M:%S")
+
+            if current_end < next_start:
+                return (current_end, next_start)
+
+        # If no free slot found, return the end of the last lease
+        last_end = datetime.strptime(node_leases[-1]['end_date'], "%Y-%m-%d %H:%M:%S")
+
+        return last_end
+
+def _call_api(endpoint):
+    url = "{0}/{1}.{2}".format(RESOURCE_API_URL, endpoint, "json")
+    LOG.info("Requesting %s from reference API ...", url)
+    resp = requests.get(url)
+    LOG.info("Response received. Parsing to json ...")
+    data = resp.json()
+    return data
+
+def get_nodes(
+        all_sites: bool = False,
+        filter_reserved: bool = False,
+        gpu: Optional[bool] = None,
+        min_number_cpu: Optional[int] = None,
+    ) -> List[Node]:
+    """
+    Retrieve a list of nodes based on the specified criteria.
+
+    Args:
+        all_sites (bool, optional): Flag to indicate whether to retrieve nodes from all sites.
+            Defaults to False.
+        filter_reserved (bool, optional): Flag to indicate whether to filter out reserved nodes.
```

Comment:
To implement this filter, you should be able to join the host data to `list_allocations` data https://docs.openstack.org/api-ref/reservation/v1/index.html#list-allocations